// vuexpress/demo/webpack.conf.js
var webpack = require('webpack');
var path = require('path');

module.exports = {
  entry: [
    'webpack-hot-middleware/client?reload=true',
    './client/app.js'
  ],
  output: {
    path: require('path').join(__dirname, 'output'),
    filename: 'app.xxxxxxx.js',
    publicPath: '/'
  },
  devtool: '#eval',
  resolve: {
    extensions: ['', '.js', '.styl']
  },
  module: {
    loaders: [
      {
        test: /\.vue$/,
        exclude: /node_modules/,
        loader: 'vue-loader'
      },
      {
        test: /\.js$/,
        exclude: /node_modules/,
        loader: 'babel'
      }
    ]
  },
  plugins: [
    new webpack.DefinePlugin({
      'process.env': {
        NODE_ENV: '"development"'
      }
    }),
    new webpack.HotModuleReplacementPlugin()
  ],
  babel: {
    presets: ['es2015', 'stage-2'],
    plugins: ['transform-runtime']
  },
  stylus: {
    import: [path.join(__dirname, 'node_modules/vuexpress/client/lib/style/mixins.styl')]
  },
  vue: {
    loaders: {
      stylus: 'style!css!stylus'
    }
  }
};
